CMakeLists.txt

```diff
@@ -0,0 +1,159 @@
+  if (NOT DEFLATE_DECOMPRESSION_ONLY)
+    list(APPEND LIBDEFLATE_SOURCES lib/gzip_compress.c)
+  endif()
+endif()
+
+set(LIBDEFLATE_HEADERS
+  common/compiler_msc.h
+  common/compiler_gcc.h
+  common/common_defs.h
+  lib/bt_matchfinder.h
+  lib/matchfinder_common.h
+  lib/unaligned.h
+  lib/gzip_constants.h
+  lib/zlib_constants.h
+  lib/crc32_vec_template.h
+  lib/decompress_template.h
+  lib/deflate_compress.h
+  lib/crc32_table.h
+  lib/x86/decompress_impl.h
+  lib/x86/crc32_impl.h
+  lib/x86/cpu_features.h
+  lib/x86/adler32_impl.h
+  lib/x86/matchfinder_impl.h
+  lib/x86/crc32_pclmul_template.h
+  lib/lib_common.h
+  lib/hc_matchfinder.h
+  lib/cpu_features_common.h
+  lib/arm/crc32_impl.h
+  lib/arm/cpu_features.h
+  lib/arm/adler32_impl.h
+  lib/arm/matchfinder_impl.h
+  lib/deflate_constants.h
+  lib/adler32_vec_template.h
+)
+
+add_library(deflate)
+add_library(deflate::deflate ALIAS deflate)
+
+set_target_properties(deflate PROPERTIES VERSION 0.0.0 SOVERSION 0)
+
+target_sources(deflate PRIVATE ${LIBDEFLATE_SOURCES}
+                               ${LIBDEFLATE_HEADERS})
+
+# Compile options
+check_c_compiler_flag(-Wall DEFLATE_LINT_WALL)
+check_c_compiler_flag(-Wundef DEFLATE_LINT_WUNDEF)
+check_c_compiler_flag(-Wpedantic DEFLATE_LINT_PEDANTIC)
+check_c_compiler_flag(-Wdeclaration-after-statement DEFLATE_LINT_DECLARATION_AFTER_STATEMENT)
+check_c_compiler_flag(-Wmissing-prototypes DEFLATE_LINT_MISSING_PROTOTYPES)
+check_c_compiler_flag(-Wstrict-prototypes DEFLATE_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wvla DEFLATE_LINT_VLA)
+check_c_compiler_flag(-Wimplicit-fallthrough DEFLATE_LINT_IMPLICIT_FALLTHROUGH)
+
+set(lint-all $<$<BOOL:${DEFLATE_LINT_WALL}>:-Wall>)
+set(lint-undef $<$<BOOL:${DEFLATE_LINT_UNDEF}>:-Wundef>)
+set(lint-pedantic $<$<BOOL:${DEFLATE_LINT_PEDANTIC}>:-Wpedantic>)
+set(lint-declaration-after-statement $<$<BOOL:${DEFLATE_LINT_DECLARATION_AFTER_STATEMENT}>:-Wdeclaration-after-statement>)
+set(lint-missing-prototypes $<$<BOOL:${DEFLATE_LINT_MISSING_PROTOTYPES}>:-Wmissing-prototypes>)
+set(lint-strict-prototypes $<$<BOOL:${DEFLATE_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
+set(lint-vla $<$<BOOL:${DEFLATE_LINT_VLA}>:-Wvla>)
+set(lint-implicit-fallthrough $<$<BOOL:${DEFLATE_LINT_IMPLICIT_FALLTHROUGH}>:-Wimplicit-fallthrough>)
+
+target_compile_options(deflate PRIVATE ${lint-all})
+target_compile_options(deflate PRIVATE ${lint-undef})
+target_compile_options(deflate PRIVATE ${lint-pedantic})
+target_compile_options(deflate PRIVATE ${lint-declaration-after-statement})
+target_compile_options(deflate PRIVATE ${lint-missing-prototypes})
+target_compile_options(deflate PRIVATE ${lint-strict-prototypes})
+target_compile_options(deflate PRIVATE ${lint-vla})
+target_compile_options(deflate PRIVATE ${lint-implicit-fallthrough})
+
+# The CFLAGS environment variable is read and handled automatically by cmake
+
+if (BUILD_SHARED_LIBS)
+  target_compile_options(deflate PRIVATE -fvisibility=hidden -D_ANSI_SOURCE)
```

Comment:
There is an abstraction of visibility preset in CMake: https://cmake.org/cmake/help/latest/prop_tgt/LANG_VISIBILITY_PRESET.html

Reply:
Moreover `LIBDEFLATE_DLL` PUBLIC definition should be added if WIN32: https://github.com/ebiggers/libdeflate/blob/e1803dc145aaecd531fb081de2b9e354fe27fda7/Makefile.msc#L46